Laboratory Work #1, Vyacheslav Sviridov - 20FPL1

VSvir opened the pull request.

VSvir committed:
Add realisation of validate_config function

VSvir committed:
Add realisation of prepare_environment function

VSvir committed:
Initialised Crawler class

VSvir committed:
Add realisation of Crawler.find_articles function

VSvir committed:
Introduce ArticleParser abstraction
Changed seed URLs

VSvir committed:
Fixed validate_config for Scrapper class
Created empty functions for ArticleParser class
Created HTMLParser class

VSvir committed:
Fixed find_articles function for Scrapper class

VSvir committed:
Made find_articles function find links for Scrapper class

marina-kaz commented on scrapper.py:
```diff
@@ -35,7 +50,25 @@ def find_articles(self):
+                print('request failed')
+
+            soup = BeautifulSoup(response.text, 'lxml')
+            main_block_bs = soup.find('div', {'class': 'two_thirds'})
```
Leave extracting links from soup object to `_extract_url`, that what this method is for

VSvir replied:
Done

marina-kaz commented on scrapper.py:
```diff
@@ -35,7 +50,25 @@ def find_articles(self):
+            main_block_bs = soup.find('div', {'class': 'two_thirds'})
+            urls = main_block_bs.find_all('a')
+            for url in urls:
+                self.urls.append('http://journals.tsu.ru' + url['href'])
```
before adding a url to urls, make sure you have not yet done this during one of the previous iterations
many sources place the same urls several times, and we want to collect unique ones

VSvir replied:
Added a check for this

marina-kaz commented on scrapper.py:
```diff
@@ -44,18 +77,63 @@ def get_search_urls(self):
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    try:
```
this will work
however, is there a way to implement this without try/except?

VSvir replied:
Reworked this function using shutil

marina-kaz commented on scrapper.py:
```diff
@@ -44,18 +77,63 @@ def get_search_urls(self):
+        if not url_validation:
+            raise IncorrectURLError
+
+    if not isinstance(max_articles, int) or max_articles <= 0:
```
this is fragile
imagine max_articles is in fact a string
then you program will crash as you are going to try to compare it to integer
it is better to separate these checks

VSvir replied:
Done

marina-kaz reviewed (changes requested):
There is not much time left so I advise to speed up a little :)

dmitry-uraev committed:
update

VSvir committed:
Changes due to comments

dmitry-uraev committed:
Merge branch 'fipl-hse:main' into main

VSvir committed:
Some changes

VSvir committed:
Merge remote-tracking branch 'origin/main'

marina-kaz commented on scrapper.py:
```diff
@@ -44,18 +77,67 @@ def get_search_urls(self):
+            shutil.rmtree(base_path)
+            base_path.mkdir(parents=True)
+    else:
+        base_path.mkdir(parents=True)
```
there is a way that is a bit simpler

marina-kaz reviewed (changes requested):
Good job, keep going!

marina-kaz commented on requirements.txt:
```diff
@@ -1 +1,3 @@
-
+beautifulsoup4==4.10.0
+pytest==7.1.0
```
you do not need pytest to run scrapper 
all the required dependencies for testing are located in requirements_qa.txt

VSvir replied:
Removed

marina-kaz reviewed (changes requested):
Just a few things to fix

marina-kaz reviewed (changes requested):
Time to finish your Crawler and embark on Pipeline :)

VSvir committed:
Add realisation of HTMLParser parse() and _fill_article_with_text() methods
changed prepare_environment() function

marina-kaz commented on scrapper.py:
```diff
@@ -44,18 +77,72 @@ def get_search_urls(self):
         pass
```
according to docstring, this method is supposed to return seed_urls value

VSvir replied:
Done it, if I understood it correctly

marina-kaz reviewed (changes requested):
Fill in `main()` and you are almost all set

VSvir committed:
Add realisation of Scrapper for mark 4

VSvir committed:
Changes according to comments + style fixes

VSvir commented on scrapper.py:
```diff
@@ -2,40 +2,73 @@
 
+import json
+import re
+import requests
```
Replaced

VSvir committed:
Add realisation of _fill_article_with_meta_information for HTMLParser class

VSvir committed:
Changed target score for scrapper

VSvir committed:
Style fixed

marina-kaz commented on scrapper.py:
```diff
@@ -141,10 +157,26 @@ def validate_config(crawler_path):
+            article = article_parser.parse()
+            article.save_raw()
+            print(f'Article {article_parser.article_id} of {number_of_articles} is done.')
+            if article_parser.article_id == number_of_articles:
```
do you expect that `crawler.urls` may contain more articles than needed?

VSvir replied:
Yes, I want to save and parse exactly as many articles, as it is specified in _config_ file

marina-kaz replied:
It is what is expected from you, right. However, judging by your `Crawler.find_articles`, crawler.urls cannot contain more articles than needed, so why perform extra checks? 
Let me know if I miss something

VSvir replied:
Yes, that's right. I forgot about it. Replaced with another check

marina-kaz reviewed (changes requested):
Nice, time to finish Crawler and start implementing Pipeline

dmitry-uraev committed:
Merge branch 'fipl-hse:main' into main

VSvir committed:
Add realisation of extracting date from meta information
Add more functions for extracting meta information
Update requirements
Target score for scrapper changed

VSvir committed:
Add seed urls
Changed needed number of articles

VSvir commented:
Can I update tests somehow? I don't think that I can pass the `s_2_5_raw_metadata_test` as it checks the html page content and it can't get it without headers

marina-kaz commented:
> Can I update tests somehow? I don't think that I can pass the `s_2_5_raw_metadata_test` as it checks the html page content and it can't get it without headers

Of course! 
Tests perform default requests with no extra headers and/or encoding settings, if you need them -- add them. But no more than that :)

marina-kaz commented on scrapper.py:
```diff
@@ -2,62 +2,232 @@
+        authors_list = []
+        for table_raw in authors_table_raws_bs:
+            authors_list.append(table_raw.find('td').text)
+        return ', '.join(authors_list)
```
According to docs, in case there are several authors, you are expected to store their names as a list of strings.

VSvir replied:
Done

marina-kaz reviewed (changes requested):
Good job!
I appreciate division of responsibility between different methods 
Time to embark on Pipeline

VSvir committed:
Add a new error class for the case when program can't find the needed number of articles
Add a new check
Implement saving of a list of authors for cases when there are several authors
Add headers for s2_5_raw_metadata_test

VSvir committed:
Code style fixed

VSvir committed:
Add more headers to s2_5_raw_metadata_test
Changed one seed_url (didn't want to deal with <i> tag in text)

VSvir committed:
Add realisation of validate_dataset function for pipeline

VSvir committed:
Add realisation of CorpusManager abstraction
Add files closing for check_txt_files function
Main function update

VSvir committed:
Add realisation of MorphologicalToken abstraction

VSvir committed:
Add realisation of TextProcessingPipeline abstraction

VSvir committed:
Implement simplified logic of TextProcessingPipeline

VSvir commented:
Should we leave English words while processing or should we better get rid of them as well?

VSvir committed:
Add realisation of Pipeline run function
Updated Pipeline _process function
Updated dataset numeration check
Add realisation of text processing program

Corrected misspellings in tests 3_4 and 3_5 descriptions

Changed target score for pipeline

VSvir committed:
Update validation check for pipeline
Add a decorator for test in s3_1 test

VSvir committed:
Changed empty files check

VSvir committed:
Trying without this check

VSvir committed:
Trying without this check (2)

VSvir committed:
Returned the check

VSvir committed:
Changed the check for imbalanced dataset

VSvir committed:
Trying another variant of dataset validation

VSvir committed:
Returned previous validation variant
Removed the decorator

VSvir committed:
Reduced the number of articles

VSvir committed:
Add realisation of MorphologicalToken get_single_tagged function
Add realisation of text processing with mystem tagging
Updated Pipeline run function
Changed target score

VSvir commented:
By the way, can I implement passing a score or a string to the `TextProcessingPipeline.run()` function so I could choose what kinds of processed text files I'd like to get?

VSvir commented:
> By the way, can I implement passing a score or a string to the `TextProcessingPipeline.run()` function so I could choose what kinds of processed text files I'd like to get?

Ah, I forgot that in that case the tests most likely won't work

marina-kaz commented:
> > By the way, can I implement passing a score or a string to the `TextProcessingPipeline.run()` function so I could choose what kinds of processed text files I'd like to get?
> 
> Ah, I forgot that in that case the tests most likely won't work

Indeed. However, if by the time you are finished with the given instructions you still feel enthusiastic about enhancing your current pipeline with extra functionality, you are very welcome to do so in a separate module. This kind of work is always appreciated, yet it must not interfere with you completing what is mandatory.

VSvir committed:
Processing raw text with mystem

marina-kaz commented on pipeline.py:
```diff
@@ -24,19 +31,22 @@ class MorphologicalToken:
+        self.original_word = original_word
+        self.normalized_form = None
+        self.tags_mystem = None
+        self.tags_pymorphy = None
```
According to the section 3 of pipeline documentation, you are expected to initialize fields of MorphologicalToken instance with empty strings

VSvir replied:
Got it

VSvir commented:
The code execution time was more than 20 minutes, 1 article. It seems like `mystem` just gets stuck while processing a file

marina-kaz commented on pipeline.py:
```diff
@@ -51,19 +61,29 @@ class CorpusManager:
         """
-        pass
+        dataset = list(self.path.glob('*_raw.txt'))
+        dataset.sort(key=self._extract_file_id)
```
could you please explain why it is important to sort files before iterating over them with the purpose of registering present articles

VSvir replied:
To be honest, I'm not sure whether it is really vital as id of the file is either way extracted in `_extract_file_id` function and `Article` instance will get the id of the processing file from its name. In my understanding, it is just safer and more comfortable to process articles in the same order as they are in the folder as `glob()` retrieves files in another order just because they are sorted not as digits but as strings if I understood it correctly. 
Maybe I don't see something?

marina-kaz reviewed (changes requested):
Good job!
Creative checks in dataset validation are appreciated

VSvir committed:
Add text preprocessing for mystem analysis

VSvir committed:
Microchanges

VSvir committed:
Add realisation of multiple tagging
Update target score

VSvir committed:
_process function update

VSvir committed:
Improved dataset validation for pipeline
Add realisation of POSFrequencyPipeline
Update target score

VSvir committed:
Dataset validation fixed

VSvir committed:
Dataset validation fixed 2.0
